File: vane-regions/src/main/java/org/oddlama/vane/regions/event/RegionEnvironmentSettingEnforcer.java
package org.oddlama.vane.regions.event;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBurnEvent;
import org.bukkit.event.block.BlockExplodeEvent;
import org.bukkit.event.block.BlockSpreadEvent;
import org.bukkit.event.entity.CreatureSpawnEvent;
import org.bukkit.event.entity.EntityChangeBlockEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityExplodeEvent;
import org.bukkit.event.entity.PotionSplashEvent;
import org.bukkit.event.hanging.HangingBreakEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.oddlama.vane.core.Listener;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.regions.Regions;
import org.oddlama.vane.regions.region.EnvironmentSetting;

public class RegionEnvironmentSettingEnforcer extends Listener<Regions> {

    public RegionEnvironmentSettingEnforcer(Context<Regions> context) {
        super(context);
    }

    public boolean checkSettingAt(
        final Location location,
        final EnvironmentSetting setting,
        final boolean checkAgainst
    ) {
        final var region = getModule().regionAt(location);
        if (region == null) {
            return false;
        }

        final var group = region.regionGroup(getModule());
        return group.getSetting(setting) == checkAgainst;
    }

    public boolean checkSettingAt(final Block block, final EnvironmentSetting setting, final boolean checkAgainst) {
        final var region = getModule().regionAt(block);
        if (region == null) {
            return false;
        }

        final var group = region.regionGroup(getModule());
        return group.getSetting(setting) == checkAgainst;
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onBlockExplode(final BlockExplodeEvent event) {
        // Prevent explosions from removing region blocks
        event.blockList().removeIf(block -> checkSettingAt(block, EnvironmentSetting.EXPLOSIONS, false));
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onEntityExplode(final EntityExplodeEvent event) {
        // Prevent explosions from removing region blocks
        event.blockList().removeIf(block -> checkSettingAt(block, EnvironmentSetting.EXPLOSIONS, false));
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onEntityChangeBlock(final EntityChangeBlockEvent event) {
        if (!(event.getEntity() instanceof Monster)) {
            return;
        }

        // Prevent monster entities from changing region blocks
        if (checkSettingAt(event.getBlock(), EnvironmentSetting.MONSTERS, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onBlockBurn(final BlockBurnEvent event) {
        if (checkSettingAt(event.getBlock(), EnvironmentSetting.FIRE, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onBlockSpread(final BlockSpreadEvent event) {
        EnvironmentSetting setting;
        switch (event.getNewState().getType()) {
            default:
                return;
            case FIRE:
                setting = EnvironmentSetting.FIRE;
                break;
            case VINE:
                setting = EnvironmentSetting.VINE_GROWTH;
                break;
        }

        if (checkSettingAt(event.getBlock(), setting, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onCreatureSpawn(final CreatureSpawnEvent event) {
        // Only cancel natural spawns and alike
        switch (event.getSpawnReason()) {
            case JOCKEY:
            case MOUNT:
            case NATURAL:
                break;
            default:
                return;
        }

        final var entity = event.getEntity();
        if (entity instanceof Monster) {
            if (checkSettingAt(event.getLocation(), EnvironmentSetting.MONSTERS, false)) {
                event.setCancelled(true);
            }
        } else if (entity instanceof Animals) {
            if (checkSettingAt(event.getLocation(), EnvironmentSetting.ANIMALS, false)) {
                event.setCancelled(true);
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onEntityDamageByEntity(final EntityDamageByEntityEvent event) {
        final var damaged = event.getEntity();
        final var damager = event.getDamager();

        if (damaged.getType() != EntityType.PLAYER) {
            return;
        }

        final Player playerDamaged = (Player) damaged;
        final Player playerDamager;
        if (damager instanceof Player) {
            playerDamager = (Player) damager;
        } else if (damager instanceof Projectile && ((Projectile) damager).getShooter() instanceof Player) {
            playerDamager = (Player) ((Projectile) damager).getShooter();
        } else {
            return;
        }

        if (
            playerDamager != null &&
            playerDamaged != playerDamager &&
            (checkSettingAt(playerDamaged.getLocation(), EnvironmentSetting.PVP, false) ||
                checkSettingAt(playerDamager.getLocation(), EnvironmentSetting.PVP, false))
        ) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onHangingBreakEvent(final HangingBreakEvent event) {
        switch (event.getCause()) {
            default:
                return;
            case ENTITY:
                return; // Handeled by onHangingBreakByEntity
            case EXPLOSION: {
                if (checkSettingAt(event.getEntity().getLocation(), EnvironmentSetting.EXPLOSIONS, false)) {
                    event.setCancelled(true);
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerInteract(final PlayerInteractEvent event) {
        if (event.getAction() != Action.PHYSICAL) {
            return;
        }

        final var block = event.getClickedBlock();
        if (block != null && block.getType() == Material.FARMLAND) {
            if (checkSettingAt(block, EnvironmentSetting.TRAMPLE, false)) {
                event.setCancelled(true);
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPotionSplash(final PotionSplashEvent event) {
        // Only if a player threw the potion check for PVP
        if (!(event.getEntity().getShooter() instanceof Player)) {
            return;
        }

        final var thrower = (Player) event.getEntity().getShooter();
        final var sourcePvpRestricted = checkSettingAt(thrower.getLocation(), EnvironmentSetting.PVP, false);

        // Cancel all damage to players if either thrower or damaged is
        // inside no-PVP region
        for (final var target : event.getAffectedEntities()) {
            if (!(target instanceof Player)) {
                continue;
            }

            if (sourcePvpRestricted || checkSettingAt(target.getLocation(), EnvironmentSetting.PVP, false)) {
                event.setIntensity(target, 0);
                return;
            }
        }
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/event/RegionEnvironmentSettingEnforcer.kt
package org.oddlama.vane.regions.event

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBurnEvent
import org.bukkit.event.block.BlockExplodeEvent
import org.bukkit.event.block.BlockSpreadEvent
import org.bukkit.event.entity.*
import org.bukkit.event.hanging.HangingBreakEvent
import org.bukkit.event.hanging.HangingBreakEvent.RemoveCause
import org.bukkit.event.player.PlayerInteractEvent
import org.oddlama.vane.core.Listener
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.regions.Regions
import org.oddlama.vane.regions.region.EnvironmentSetting

/**
 * Enforces environment-related region settings such as PVP, fire, and mob spawning.
 */
class RegionEnvironmentSettingEnforcer(context: Context<Regions?>?) : Listener<Regions?>(context) {
    /**
     * Owning regions module instance.
     */
    private val regions: Regions
        get() = requireNotNull(module)

    /**
     * Checks whether an environment setting at a location matches the expected value.
     */
    fun checkSettingAt(
        location: Location,
        setting: EnvironmentSetting,
        checkAgainst: Boolean
    ): Boolean {
        /** Region at the queried location. */
        val region = regions.regionAt(location) ?: return false
        /** Region group owning that region. */
        val group = region.regionGroup(regions) ?: return false
        return group.getSetting(setting) == checkAgainst
    }

    /**
     * Checks whether an environment setting at a block matches the expected value.
     */
    fun checkSettingAt(block: Block, setting: EnvironmentSetting, checkAgainst: Boolean): Boolean {
        val region = regions.regionAt(block) ?: return false
        val group = region.regionGroup(regions) ?: return false
        return group.getSetting(setting) == checkAgainst
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Prevents explosion block damage where `EXPLOSIONS` is denied.
     */
    fun onBlockExplode(event: BlockExplodeEvent) {
        // Prevent explosions from removing region blocks
        event.blockList().removeIf { block -> checkSettingAt(block, EnvironmentSetting.EXPLOSIONS, false) }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Prevents entity-caused explosion block damage where `EXPLOSIONS` is denied.
     */
    fun onEntityExplode(event: EntityExplodeEvent) {
        // Prevent explosions from removing region blocks
        event.blockList().removeIf { block -> checkSettingAt(block, EnvironmentSetting.EXPLOSIONS, false) }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Prevents monster block changes where monster interactions are denied.
     */
    fun onEntityChangeBlock(event: EntityChangeBlockEvent) {
        if (event.entity !is Monster) {
            return
        }

        // Prevent monster entities from changing region blocks
        if (checkSettingAt(event.block, EnvironmentSetting.MONSTERS, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Cancels burning where `FIRE` is denied.
     */
    fun onBlockBurn(event: BlockBurnEvent) {
        if (checkSettingAt(event.block, EnvironmentSetting.FIRE, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Cancels fire/vine spread where corresponding settings are denied.
     */
    fun onBlockSpread(event: BlockSpreadEvent) {
        val setting = when (event.newState.type) {
            Material.FIRE -> EnvironmentSetting.FIRE
            Material.VINE -> EnvironmentSetting.VINE_GROWTH
            else -> return
        }

        if (checkSettingAt(event.block, setting, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Restricts natural animal/monster spawns based on region settings.
     */
    fun onCreatureSpawn(event: CreatureSpawnEvent) {
        // Only cancel natural spawns and alike
        when (event.spawnReason) {
            CreatureSpawnEvent.SpawnReason.JOCKEY, CreatureSpawnEvent.SpawnReason.MOUNT, CreatureSpawnEvent.SpawnReason.NATURAL -> {}
            else -> return
        }

        val entity = event.entity
        if (entity is Monster) {
            if (checkSettingAt(event.location, EnvironmentSetting.MONSTERS, false)) {
                event.isCancelled = true
            }
        } else if (entity is Animals) {
            if (checkSettingAt(event.location, EnvironmentSetting.ANIMALS, false)) {
                event.isCancelled = true
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Enforces PVP restrictions for direct and projectile player damage.
     */
    fun onEntityDamageByEntity(event: EntityDamageByEntityEvent) {
        val damaged = event.getEntity()
        val damager = event.damager

        if (damaged.type != EntityType.PLAYER) {
            return
        }

        val playerDamaged = damaged as Player
        val playerDamager = when (damager) {
            is Player -> {
                damager
            }

            is Projectile if damager.shooter is Player -> {
                damager.shooter as Player?
            }

            else -> {
                return
            }
        }

        if (playerDamager != null && playerDamaged !== playerDamager &&
            (checkSettingAt(playerDamaged.location, EnvironmentSetting.PVP, false) ||
                    checkSettingAt(playerDamager.location, EnvironmentSetting.PVP, false))
        ) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Prevents explosion-based hanging breaks where explosions are denied.
     */
    fun onHangingBreakEvent(event: HangingBreakEvent) {
        when (event.cause) {
            RemoveCause.ENTITY -> return  // Handeled by onHangingBreakByEntity
            RemoveCause.EXPLOSION -> {
                if (checkSettingAt(event.entity.location, EnvironmentSetting.EXPLOSIONS, false)) {
                    event.isCancelled = true
                }
            }

            else -> return
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Prevents farmland trampling when `TRAMPLE` is denied.
     */
    fun onPlayerInteract(event: PlayerInteractEvent) {
        if (event.action != Action.PHYSICAL) {
            return
        }

        val block = event.clickedBlock
        if (block != null && block.type == Material.FARMLAND) {
            if (checkSettingAt(block, EnvironmentSetting.TRAMPLE, false)) {
                event.isCancelled = true
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    /**
     * Enforces PVP restrictions for splash-potion effects.
     */
    fun onPotionSplash(event: PotionSplashEvent) {
        // Only if a player threw the potion check for PVP
        val thrower = event.entity.shooter as? Player ?: return
        val sourcePvpRestricted = checkSettingAt(thrower.location, EnvironmentSetting.PVP, false)

        // Cancel all damage to players if either thrower or damaged is
        // inside no-PVP region
        for (target in event.affectedEntities) {
            if (target !is Player) {
                continue
            }

            if (sourcePvpRestricted || checkSettingAt(target.location, EnvironmentSetting.PVP, false)) {
                event.setIntensity(target, 0.0)
                return
            }
        }
    }
}
